crates/fuzzing/src/generators/component_types.rs

```diff
@@ -0,0 +1,858 @@
+            Self::F64 => f.write_str("f64"),
+        }
+    }
+}
+
+/// Wraps a `Box<[T]>` and provides an `Arbitrary` implementation that always generates non-empty slices
+#[derive(Debug)]
+pub struct NonEmptyArray<T>(Box<[T]>);
+
+impl<'a, T: Arbitrary<'a>> Arbitrary<'a> for NonEmptyArray<T> {
+    fn arbitrary(input: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        Ok(Self(
+            iter::once(input.arbitrary())
+                .chain(input.arbitrary_iter()?)
+                .collect::<arbitrary::Result<_>>()?,
+        ))
+    }
+}
+
+/// Wraps a `Box<[T]>` and provides an `Arbitrary` implementation that always generates non-empty slices of length
+/// less than or equal to the longest tuple for which Wasmtime generates a `ComponentType` impl
+#[derive(Debug)]
+pub struct TupleArray<T>(Box<[T]>);
+
+impl<'a, T: Arbitrary<'a>> Arbitrary<'a> for TupleArray<T> {
+    fn arbitrary(input: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        Ok(Self(
+            iter::once(input.arbitrary())
+                .chain(input.arbitrary_iter()?)
+                .take(MAX_TUPLE_LENGTH)
+                .collect::<arbitrary::Result<_>>()?,
+        ))
+    }
+}
+
+/// Represents a component model interface type
+#[allow(missing_docs)]
+#[derive(Arbitrary, Debug)]
+pub enum Type {
+    Unit,
+    Bool,
+    S8,
+    U8,
+    S16,
+    U16,
+    S32,
+    U32,
+    S64,
+    U64,
+    Float32,
+    Float64,
+    Char,
+    String,
+    List(Box<Type>),
+    Record(NonEmptyArray<Type>),
+    Tuple(TupleArray<Type>),
+    Variant(NonEmptyArray<Type>),
+    Enum(NonEmptyArray<()>),
+    Union(NonEmptyArray<Type>),
+    Option(Box<Type>),
+    Expected { ok: Box<Type>, err: Box<Type> },
+    Flags(NonEmptyArray<()>),
+}
+
+impl fmt::Display for Type {
+    /// Format this type according to [the component model
+    /// grammar](https://github.com/WebAssembly/component-model/blob/main/design/mvp/Explainer.md#type-definitions).
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Unit => f.write_str("unit"),
+            Self::Bool => f.write_str("bool"),
+            Self::S8 => f.write_str("s8"),
+            Self::U8 => f.write_str("u8"),
+            Self::S16 => f.write_str("s16"),
+            Self::U16 => f.write_str("u16"),
+            Self::S32 => f.write_str("s32"),
+            Self::U32 => f.write_str("u32"),
+            Self::S64 => f.write_str("s64"),
+            Self::U64 => f.write_str("u64"),
+            Self::Float32 => f.write_str("float32"),
+            Self::Float64 => f.write_str("float64"),
+            Self::Char => f.write_str("char"),
+            Self::String => f.write_str("string"),
+            Self::List(ty) => write!(f, "(list {ty})"),
+            Self::Record(types) => {
+                f.write_str("(record")?;
+                for (index, ty) in types.0.iter().enumerate() {
+                    write!(f, r#" (field "f{index}" {ty})"#)?;
```

Comment:
Like you mentioned in the issue this can lead to too-deep recursion in the text format. I think a better strategy might be to generate a `(type $name (record ...))` for each layer of a `Type` perhaps? That would avoid the `*.wast` limitations while still helping to fuzz super-deep type hierarchies in Wasmtime (which I think would actually be quite useful, so I'd rescind my other suggestion of limiting the depth here)

Reply:
Oops, I didn't see this until after I wrote https://github.com/bytecodealliance/wasmtime/pull/4537/commits/8fceba5be1f00c72582b5deaf6b490e421dfc887.  Should I undo that and use this approach instead?

Reply:
Ah sorry for giving conflicting advice! I think it would be best to exercise as interesting type hierarchies as we can (including weirdly deep ones) so I think it'd be better to skip the `*.wast`-based limits. That being said if you'd prefer to get something working first I'm definitely ok having a follow-up issue for this.

Reply:
Ugh, https://github.com/bytecodealliance/wasmtime/commit/8fceba5be1f00c72582b5deaf6b490e421dfc887 didn't really address the issue anyway, so I'll try your approach.